packages/docgen/test/get-type-annotation.js

```diff
@@ -415,6 +415,68 @@ describe( 'Type annotations', () => {
 			).toBe( 'string[]' );
 		} );
 
+		it( 'should lean on `getDependants` if `selector` in `createSelector` is missing args', () => {
+			const { tokens } = engine(
+				'test.ts',
+				`/**
+				 * Returns a reference that changes only when the dependencies do.
+				 *
+				 * @param state - stores all the things
+				 */
+				 export const getCountReference = createSelector( () => [], ( state: State ) => [ state.items.length ] );
+				 `
+			);
+
+			expect(
+				getTypeAnnotation(
+					{ tag: 'param', name: 'state' },
+					tokens[ 0 ],
+					0
+				)
+			).toBe( 'State' );
```

Comment:
So just to understand, this is returning State as the argument because all selectors accept the single State argument regardless of whether they do anything?

If that's a correct description of the intention, then I think this change is more confusing than helpful from a documentation perspective. The function (selector) takes no arguments at all. Functionally it takes a state argument but it does nothing with it, but suggesting to the user that it accepts some state is incorrect, I think?

Is state passed implicitly to the selectors before the arguments passed by the callsite of the selector? It seems like in those cases you don't even need to document that the selector accepts a state argument at all, it's an implementation detail, not part of the public API.

Maybe I'm not understanding this correctly though?

Reply:
Your understanding is correct on all levels. Selector functions are supplied `state` by `core/data` and not presented to calling code; they are auto-curried with `state`.

To that end, a few more complicated situations ignored, it's wrong to create a selector whose first argument is anything other than `state` as that will break at runtime when it's provided that `state`.

We don't have to use `state` at all though and that's fine here. What the test is demonstrating is that the arguments returned by `createSelector` are the same arguments for both passed functions because it's whichever passed function that has the longest array of input arguments which determines the arguments for the returned wrapped function. You can't call `selector` here because it's wrapped in the behavior of `createSelector` and if you do so without `state` then it will probably crash when it internally calls `getDependants`.

If these functions weren't prevalent _everywhere_ in the source code I wouldn't want to manually exempt them because this is truly just a failure of not asking TypeScript what the parameters and types are. `createSelector` would likely have the parameters as some kind of type constraint to adequately communicate this here.

```ts
function createSelector<
    SelectorArgs extends any[],
    Selector extends (...args: SelectorArgs) => any
>(
    selector: Selector,
    getDependants?: (...args: SelectorArgs) => unknown[]
): ReturnType<Selector> {
    …
}
```

Here TypeScript already handles the case where `selector` lists fewer arguments because that's how they built the system. any function with fewer arguments satisfies a type of a function with more arguments if the ones it does list match the ones in the larger type.

Reply:
Thanks for clarifying. I still don't understand why state should be documented as a parameter on the public api of the selector function. It seems needlessly confusing to me. If I was reading the documentation I'd expect to have to pass the state argument despite it being curried in, as you said, and probably despite everything I knew about how "selectors" worked up until that point in the redux-y land that the data package lives in.

I don't know how to easily accommodate ignoring the state parameter. We could continue to build the exceptions to createSelector and tell it to ignore the state parameter at all times. Or maybe create some kind of documentation annotation that could be applied to parameters we want ignored from documentation. The first continues down the rather unfortunate path of docgen predicting particular quirks about WordPress code (though it's already fairly specific to the projects use case anyway) and the second adds verbosity and additional expertise required to write a properly documented selector. I'd probably prefer the first, even though it's "magic" it will stay out of the way of people just trying to write a selector.

Reply:
> why state should be documented as a parameter on the public api of the selector function

the confusion probably comes from the fact that these aren't the public selectors just as this isn't the public API. these are the underlying Redux selectors that _do_ require `state` and the precedent is documenting `state` [(for example)](https://github.com/WordPress/gutenberg/tree/trunk/packages/core-data#getlastentitydeleteerror) because that is the type of these functions.

the public API you are referring to comes as a new function produced by the data system and accessed through the `select( store )` interface.

we'd have to lie in the type system if we ignored it from these functions, and all of the same kind which are written in `selectors.js` and `selectors.ts`. on the other hand, if we wrote these functions without `state` then they would break when someone tries to use the returned versions that the core data store produces.